[seraphis] crypto: miscellaneous updates

UkoeHB opened the pull request:
This is a PR in my 'upstreaming seraphis_lib project', the changes here are not used anywhere yet.

- Make some `crypto-ops.c` static methods public. These will be used in PR #8662 for recomputing the main ed25519 generator.
- Allow sorting of key images.
- Add `to_bytes()` methods for converting `ec_point` and `ec_scalar` to unsigned char pointers. These methods are SUPER helpful when doing anything with the `sc_*` methods. Note that there is a variant of this in `device_default.cpp`.
- Rename `rct::ecdhHash()` to `rct::genAmountEncodingFactor()` and make it public.

SChernykh commented on src/crypto/crypto-ops.h:
```diff
@@ -30,6 +30,8 @@
 
 #pragma once
 
+#include <stdint.h>
```
Why was it added here? `crypto-ops.c` already includes it before `crypto.h`

SChernykh replied:
It makes sense then to remove `#include <stdint.h>` from `crypto-ops.c`

UkoeHB replied:
This cleans up static analyzer errors.

If you [google transitive includes](https://www.google.com/search?client=firefox-b-1-d&q=transitive+includes) you don't find anyone recommending they be used. Here is the somewhat authoritative [cpp core guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-implicit).

jeffro256 commented on src/crypto/crypto.h:
```diff
@@ -335,11 +335,19 @@ namespace crypto {
   inline bool operator<(const public_key &p1, const public_key &p2) { return memcmp(&p1, &p2, sizeof(public_key)) < 0; }
   inline bool operator>(const public_key &p1, const public_key &p2) { return p2 < p1; }
+  inline bool operator<(const key_image &p1, const key_image &p2) { return memcmp(&p1, &p2, sizeof(key_image)) < 0; }
+  inline bool operator>(const key_image &p1, const key_image &p2) { return p2 < p1; }
```
Looks good all around, but just personal preference:
```suggestion
  inline bool operator>(const key_image &p1, const key_image &p2) { return memcmp(&p1, &p2, sizeof(key_image)) > 0; }
```
There's a 90% chance this will get optimized out by the compiler, but since this is gonna likely take place many, many times a second , it might be better to write it out explicitly.

UkoeHB replied:
The other operator is `inline`. If this is not inlined then idk what to say - compilers must be trash.

vtnerd replied:
This comment is old, but also the `inline` keyword is needed here or linker errors could occur.

jeffro256 commented on src/crypto/crypto-ops.h:
```diff
@@ -161,5 +163,12 @@ void ge_sub(ge_p1p1 *r, const ge_p3 *p, const ge_cached *q);
 
 int ge_p3_is_point_at_infinity_vartime(const ge_p3 *p);
+
+// miscellaneous
```
Since this is in an important file, I'd leave a more descriptive comment here, more or less what you put in the post, explaining why this needs to be exposed in this header file.

UkoeHB replied:
Oh looks like these are leftovers I'm not actually using. Will remove

jeffro256 reviewed (commented):
Looks good to me

vtnerd commented on src/crypto/crypto.h:
```diff
@@ -335,8 +335,16 @@ namespace crypto {
 }
 
+// type conversions for easier calls to sc_add(), sc_sub(), hash functions
+inline unsigned char* to_bytes(crypto::ec_scalar &scalar) { return &reinterpret_cast<unsigned char&>(scalar); }
```
These should be dropped in favor of `epee::to_byte_span()` usage. Or just redirect to that function and return a `span` here, although I don't see a purpose of re-creating those existing functions.

UkoeHB replied:
If I use the span stuff directly, then:
```
sc_add(to_bytes(a), to_bytes(b), to_bytes(c));  //old
sc_add(epee::as_mut_byte_span(a).data(), epee::as_byte_span(b).data(), epee::as_byte_span(c).data());  //new
```

Using a `span` internally means you have to construct a `span` every time `to_bytes` is called, which is not ideal for one of the main use cases - high performance crypto algorithms.

vtnerd replied:
The compiler should optimize away all references to the span if only the pointer is used. This is pretty basic stuff for the compiler actually.

I wasn't aware of the context though - these existing functions take a pointer with no other context anyway. My frustration with adding these functions is the the temptation to use these in non-crypto related functions. I guess reviewers will have to catch this.

UkoeHB replied:
It's hard for me to imagine a case where `to_bytes()` would be used for anything other than exposing a raw pointer to C-style crypto functions.

jeffro256 replied:
I think I fall on the side of @UkoeHB here. Most of the underlying C crypto functions already use "unsafe" pointers insofar as the size of the area passed is implied by the contract of the function, not enforced at compile-time or runtime. Unless we want to rewrite these functions, using `epee::span` would just add an unused wrapper on top of a pointer which may or may not get optimized out by the compiler.

UkoeHB commented:
@vtnerd I squashed commits and added one thing. I need the method `rct::ecdhHash()`, so I renamed to `genAmountEncodingFactor()` and made it public. Better to shove that in this PR instead of making a whole new PR.
